cli/replicaset/cconfig.go

```diff
@@ -204,3 +219,411 @@ func updateCConfigInstances(replicaset *Replicaset, topology cconfigTopology) {
+	if err != nil {
+		return fmt.Errorf("failed to collect cluster config: %w", err)
+	}
+
+	clusterConfig, err := libcluster.MakeClusterConfig(config)
+	if err != nil {
+		return fmt.Errorf("failed to make cluster config: %w", err)
+	}
+
+	targetGroup, targetReplicaset, found := cconfigFindInstance(&clusterConfig, instName)
+	if !found {
+		return fmt.Errorf("instance %q not found in the cluster configuration", instName)
+	}
+
+	failover, err := cconfigGetFailover(&clusterConfig, instName)
+	if err != nil {
+		return err
+	}
+
+	var (
+		targetPath []string
+		lowerDepth int
+	)
+	switch failover {
+	case FailoverOff:
+		targetPath = []string{"groups", targetGroup, "replicasets",
+			targetReplicaset, "instances", instName, "database", "mode"}
+		lowerDepth = len(targetPath) - 2
+	case FailoverManual:
+		targetPath = []string{"groups", targetGroup, "replicasets",
+			targetReplicaset, "leader"}
+		lowerDepth = len(targetPath) - 1
+	case FailoverElection:
+		return fmt.Errorf(`unsupported failover: %q, supported: "manual", "off"`, failover)
+	default:
+		return fmt.Errorf(`unknown failover, supported: "manual", "off"`)
+	}
+
+	configData, err := dataCollector.Collect()
+	if err != nil {
+		return fmt.Errorf("failed to collect cluster config data: %w", err)
+	}
+
+	var targets []struct {
+		key      string
+		revision int64
+		decoded  *libcluster.Config
+		priority int
+	}
+
+	for _, data := range configData {
+		config, err := libcluster.NewYamlCollector(data.Value).Collect()
+		if err != nil {
+			return fmt.Errorf("failed to decode the config by the key %q: %w", data.Source, err)
+		}
+
+		depth, err := cconfigGetDepth(config, targetPath, lowerDepth)
+		if err != nil {
+			return err
+		}
+		if depth == noDepth {
+			continue
+		}
+		targets = append(targets, struct {
+			key      string
+			revision int64
+			decoded  *libcluster.Config
+			priority int
+		}{
+			key:      data.Source,
+			revision: data.Revision,
+			decoded:  config,
+			priority: depth,
+		})
+	}
+
+	sort.Slice(targets, func(i, j int) bool {
+		ip := targets[i].priority
+		jp := targets[j].priority
+		if ip != jp {
+			return ip > jp
+		}
+		// If the priorities are equal lexicographically smaller keys are first,
+		// to make the changes have an effect.
+		return targets[i].key < targets[j].key
+	})
+
+	targetKeys := make([]string, 0, len(targets))
+	for _, target := range targets {
+		targetKeys = append(targetKeys, target.key)
+	}
+
+	dstIndex, err := pickKey(targetKeys)
+	if err != nil {
+		return err
+	}
+
+	dst := targets[dstIndex].key
+	patched, err := cconfigPatchPromote(targets[dstIndex].decoded, failover,
+		targetGroup, targetReplicaset, instName)
+	if err != nil {
+		return fmt.Errorf("failed to patch the config: %w", err)
+	}
+
+	err = c.publisher.Publish(dst, []byte(patched.String()),
+		targets[dstIndex].revision)
+	if err != nil {
+		return fmt.Errorf("failed to publish the config: %w", err)
+	}
+	return nil
+}
```

Comment:
This method does too much, at first glance it generally looks like the logic of a separate command and it is not entirely clear why it should be put into a separate type. This requires decomposition.

Reply:
>  and it is not entirely clear why it should be put into a separate type

Do you suggest to make it just function? 
We have CconfigInstance and CconfigApplication, so I thought to add a new type to do the config source manipulations with it. But, the same time CConfigSource does not implement Promoter, because it used in a different context.

Reply:
Do you plan to extend the type with new methods?
Why the `CConfigSource` could not to implement `Promote` interface. I understand that now it is use a different context, but in theory, is there are a principal problem?

Reply:
I explained new changes connected to the `CConfigSource` in the comment above. Btw, the method is decomposed now.